### DISCUSSION TITLE
Docker container action - 'right' way to equivalent a volume?

### DISCUSSION THREAD

**Discussion by GavinCameron** [community/community#25637] [Actions]
<p>I have a repo I created whose sole purpose is to create a Docker container action that connects to a database (Snowflake) and outputs some data.</p>
<p>I drew heavily <a href="https://github.com/kurron/docker-snowsql" rel="noopener nofollow ugc">from this repo</a> and was able to recreate locally in Docker.</p>
<p>Part of the set up uses a config file. Anyone in our org using this repo will need their own config file. On the repo linked above you can see this in the <a href="https://github.com/kurron/docker-snowsql/blob/b1b16ff88b36f9df58f692a69642e84f353c2d46/connection-test.sh" rel="noopener nofollow ugc">connection-test.sh script</a> where row 5 uses a custom config file.</p>
<p>I’m trying to figure out the equivalent of doing this with gh actions where I have a repo that contains the base image or action, and then in another repo I want to checkout the snowsql action and then when run use a custom config file that I have in my repo that uses the action.</p>
<ul>
<li>ourorg/snowsql # the generic docker container action repo</li>
<li>ourorg/my-project-that-uses-snowsql</li>
</ul>
<p>When I set this up just using Docker on local I have a docker-compose file of the form:</p>
<pre><code class="lang-auto">volumes:
      - ./ssql/config:/home/snowflake/.snowsql/config
</code></pre>
<p>Is there an equivalent or ‘right’ way to do this using a custom docker action?</p>
<p>Repo <code>ourorg/my-project-that-uses-snowsql.github/workflows/main.yml</code> looks like this:</p>
<pre><code class="lang-auto">on: [push]

jobs:
  ssql_job:
    runs-on: ubuntu-latest
    name: Build ssql image
    steps:
      - name: Checkout # To use this repository's private action, you must check out the repository
        uses: actions/checkout
	with: # Aside, never done this before, don't know if it's right
	  repository: ourorg/ds-ssql-gh-action
	  ref: main
      - name: Build
	uses: ./ # Uses an action in the root directory 
</code></pre>
<p>If this repo <code>ourorg/ds-ssql-gh-action</code> has a file <code>./config</code> and after I checkout <code>ourorg/ds-ssql-gh-action</code> I’d like to run the action / container but before doing so I want to add  <code>ourorg/my-project-that-uses-snowsql/config</code> inside the action / container at <code>/home/snowflake/.snowsql/config</code>. LIke with a volume mount in docker-compose.</p>
<p>How can I do this?</p>

**Comment by airtower-luna** [comment]
<p>You can’t do custom volume mounts in Docker actions. The closest thing I can think of would be to have an input that gives the (workspace-relative) path of the config file to use. Your action could then either call whatever needs the config with appropriate parameters to set the config path, or (if the config path is fixed) copy the file into place within the action container.</p>

**Comment by GavinCameron** [comment]
<p>Hi <a class="mention" href="/airtower-luna">@airtower-luna</a> could you expand on this?</p>
<blockquote>
<p>copy the file into place within the action container</p>
</blockquote>
<p>How?! I don’t understand.</p>
<p>To underline, I am checking out a generic action from another repo, then I with to augment it by copying the config file in the new repo to the action container.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>When you run a container action (e.g. <code>uses: my/action@tag</code>) the runner runs that container, mounts the job workspace into it, and makes it the working directory of the container. So the code running inside the container has access to the workspace and can copy a file from there to another location inside the container. Though it’d be preferable to just use the file as is, but if you use a 3rd party tool that uses a fixed path copying might be the easiest solution.</p>
<p>Assuming you have a file <code>my/config.conf</code> in your working directory and <code>my/action</code> defines a <code>config</code> input, you could call your action like this:</p>
<pre><code class="lang-yaml">- uses: my/action@tag
  with:
    config: my/config.conf
</code></pre>
<p>The runner defines an <code>INPUT_CONFIG</code> environment variable inside the container that contains the input value, so if your entrypoint is a shell script you could do something like this in there (assuming the tool you’re about to run requires a fixed location):</p>
<pre><code class="lang-auto">cp "${INPUT_CONFIG}" "${HOME}/.tool/config"
</code></pre>
<p>Or if the tool has a command line parameter to set the config file (better):</p>
<pre><code class="lang-auto">tool --config "${INPUT_CONFIG}"
</code></pre>
<p>One thing to keep in mind when working with the home directory is that the user and home directory within the container might not be what you expect, see <a href="https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions" class="inline-onebox">Dockerfile support for GitHub Actions - GitHub Docs</a>, so avoid it if possible. :slightly_smiling_face:</p>

**Comment by GavinCameron** [comment]
<p>This is incredibly helpful! Sincere thank you.</p>

**Reply by hridyeshpant** [reply]
Hi @GavinCameron,  i have same requirement but not able to fix it. If you can snapshot with example of action file. that will be great.

**Reply by hridyeshpant** [reply]
not sure if my case is different from yours, I posted my question in https://github.com/orgs/community/discussions/54823
